ledger/complete/mtrie/trie/trie.go

```diff
@@ -375,37 +418,59 @@ func update(
 
 	// recurse over each branch
 	var lChild, rChild *node.Node
+	var lRegCountDelta, rRegCountDelta int64
+	var lRegSizeDelta, rRegSizeDelta int64
+	var lLowestHeightTouched, rLowestHeightTouched int
 	parallelRecursionThreshold := 16
 	if len(lpaths) < parallelRecursionThreshold || len(rpaths) < parallelRecursionThreshold {
 		// runtime optimization: if there are _no_ updates for either left or right sub-tree, proceed single-threaded
-		lChild = update(nodeHeight-1, lchildParent, lpaths, lpayloads, lcompactLeaf, prune)
-		rChild = update(nodeHeight-1, rchildParent, rpaths, rpayloads, rcompactLeaf, prune)
+		lChild, lRegCountDelta, lRegSizeDelta, lLowestHeightTouched = update(nodeHeight-1, lchildParent, lpaths, lpayloads, lcompactLeaf, prune)
+		rChild, rRegCountDelta, rRegSizeDelta, rLowestHeightTouched = update(nodeHeight-1, rchildParent, rpaths, rpayloads, rcompactLeaf, prune)
 	} else {
 		// runtime optimization: process the left child is a separate thread
-		wg := sync.WaitGroup{}
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			lChild = update(nodeHeight-1, lchildParent, lpaths, lpayloads, lcompactLeaf, prune)
-		}()
-		rChild = update(nodeHeight-1, rchildParent, rpaths, rpayloads, rcompactLeaf, prune)
-		wg.Wait()
+
+		// Since we're receiving 4 items from goroutine, use a
+		// channel to prevent them going on the heap.
+		type updateResult struct {
+			child               *node.Node
+			regCountDelta       int64
+			regSizeDelta        int64
+			lowestHeightTouched int
+		}
+		results := make(chan updateResult, 1)
+		go func(retChan chan<- updateResult) {
+			child, regCountDelta, regSizeDelta, lowestHeightTouched := update(nodeHeight-1, lchildParent, lpaths, lpayloads, lcompactLeaf, prune)
+			retChan <- updateResult{child, regCountDelta, regSizeDelta, lowestHeightTouched}
+		}(results)
+
+		rChild, rRegCountDelta, rRegSizeDelta, rLowestHeightTouched = update(nodeHeight-1, rchildParent, rpaths, rpayloads, rcompactLeaf, prune)
+
+		// Wait for results from goroutine.
+		ret := <-results
+		lChild, lRegCountDelta, lRegSizeDelta, lLowestHeightTouched = ret.child, ret.regCountDelta, ret.regSizeDelta, ret.lowestHeightTouched
```

Comment:
> Have you benchmarked the construction with the channels here? This is very performance-sensitive code. 

Yes. Last week, I ran non-micro benchmarks.  Checkpoint creation benchmark using mainnet data (replays 264881 updates from WALs).  Using `channel` instead of `WaitGroup` was faster and reduced allocs by hundreds of millions.

I also ran micro benchmarks yesterday and they confirm the suggested `WaitGroup` approach can be slower than this PR by about 12%-21% while also causing extra 1-4 allocs/op depending on details (see below).

> I would guess that a waitgroup is much more performant than a channel 

Although `WaitGroup` approach can be faster (especially with 2+ goroutines), we only use 1 goroutine and need to communicate results from it.  So using `channel` is faster and uses fewer `allocs/op` in this case.
 
> ### Suggestion:
> I think this synchronization can be implemented as easily using WaitGroup:
> 
> 1. Create an instance of `updateResult` and a `wg:= sync.WaitGroup{}`
> 2. spawn the auxiliary go routine
> ...
> In summary, I think the entire else-branch can be written as follows

<img width="725" alt="image" src="https://user-images.githubusercontent.com/33205765/158677172-79f0ae46-a84f-48ac-820c-736631e80d38.png">

The suggested approach in the code snippet (screenshot) can be slower by around 21% and also cause an extra 4 allocs/op.  

The extra allocs/op (`lChild`, `lRegCountDelta`, `lRegSizeDelta`, `lLowestHeightTouched`) is incurred for all recursive calls, even for calls that use the "if-else" branch avoiding the goroutine.

> 1. Create an instance of `updateResult` and a `wg:= sync.WaitGroup{}`

If the 4 values are put into `updateResult` struct, then using the suggested `WaitGroup` approach still causes 1 extra alloc/op and it's still slower than using `channel` in this PR by around 12%.

## More Details

For micro benchmarks today, I used Go 1.17.8 on linux_amd64 (Haswell Xeon) with this code:
https://gist.github.com/fxamacker/4cec13885dcf2c19d055b13f2b51289b

There are two groups of benchmarks: 
- [WaitGroup vs channel](https://gist.github.com/fxamacker/4cec13885dcf2c19d055b13f2b51289b#file-goroutine_bench_test-go-L29-L50) (signal only, no data values communicated)
- [communicating values from goroutine](https://gist.github.com/fxamacker/4cec13885dcf2c19d055b13f2b51289b#file-goroutine_bench_test-go-L52-L109)

### WaitGroup vs channel

With one goroutine, it is faster to use channel. 

```
name         old time/op    new time/op    delta
WaitGroup-8     471ns ± 0%     448ns ± 0%    -4.79%  (p=0.000 n=20+20)

name         old alloc/op   new alloc/op   delta
WaitGroup-8     32.0B ± 0%    112.0B ± 0%  +250.00%  (p=0.000 n=20+20)

name         old allocs/op  new allocs/op  delta
WaitGroup-8      2.00 ± 0%      2.00 ± 0%      ~     (all equal)
```

### Communicating values from goroutine

Communicating 4 ints using referenced variables with WaitGroup causes 4 integers to allocate on the heap, and is also slower.

```
> go-benchrun WaitGroupReferenceFourInts ChannelResult -benchmem -count=20

name                        old time/op    new time/op    delta
WaitGroupReferenceFourInts-8     580ns ± 0%     457ns ± 0%  -21.15%  (p=0.000 n=19+20)

name                        old alloc/op   new alloc/op   delta
WaitGroupReferenceFourInts-8     96.0B ± 0%    112.0B ± 0%  +16.67%  (p=0.000 n=20+20)

name                        old allocs/op  new allocs/op  delta
WaitGroupReferenceFourInts-8      6.00 ± 0%      2.00 ± 0%  -66.67%  (p=0.000 n=20+20)
```

Communicating 1 wrapper referenced struct with WaitGroup requires allocating struct on the heap, and is slower.

```

> go-benchrun WaitGroupReferenceOneStruct ChannelResult -benchmem -count=20

name                        old time/op    new time/op    delta
WaitGroupReferenceOneStruct-8 520ns ± 0% 458ns ± 0% -11.91% (p=0.000 n=19+17)

name                        old alloc/op   new alloc/op   delta
WaitGroupReferenceOneStruct-8 72.0B ± 0% 112.0B ± 0% +55.56% (p=0.000 n=20+20)

name                        old allocs/op  new allocs/op  delta
WaitGroupReferenceOneStruct-8 3.00 ± 0% 2.00 ± 0% -33.33% (p=0.000 n=20+20)

```

Although `WaitGroup` approach can be faster (especially with 2+ goroutines), we only use 1 goroutine and need to communicate results from it.  So using `channel` is faster and uses fewer `allocs/op` in this case.

Reply:
Thanks for the detailed benchmarking result. Makes sense. Could you add a brief comment to the code, so that we document this consideration.

Reply:
Added in 051ba8e18a76f70d13d9d6f5089767b5bb839c33

Reply:
Interesting results! I didn't expect `channel` to be faster than `waitGroup` in this case.